dht.go

```diff
@@ -219,16 +223,42 @@ func NewDHTClient(ctx context.Context, h host.Host, dstore ds.Batching) *IpfsDHT
 }
 
 func makeDHT(ctx context.Context, h host.Host, cfg config) (*IpfsDHT, error) {
-	protocols := []protocol.ID{cfg.protocolPrefix + kad2}
-	serverProtocols := []protocol.ID{cfg.protocolPrefix + kad2, cfg.protocolPrefix + kad1}
+	var (
+		queryProtocols, serverProtocols []protocol.ID
+		routingTableProtocols           []string
+	)
 
 	// check if custom test protocols were set
 	if len(cfg.testProtocols) > 0 {
-		protocols = make([]protocol.ID, len(cfg.testProtocols))
-		serverProtocols = make([]protocol.ID, len(cfg.testProtocols))
+		protocols := make([]protocol.ID, len(cfg.testProtocols))
 		for i, p := range cfg.testProtocols {
 			protocols[i] = cfg.protocolPrefix + p
-			serverProtocols[i] = cfg.protocolPrefix + p
+		}
+		queryProtocols = protocols
+		routingTableProtocols = protocol.ConvertToStrings(protocols)
+		serverProtocols = protocols
+	} else {
+		if cfg.v1CompatibleMode {
+			// In compat mode, query/serve using the old protocol,
+			// but _accept_ peers that speak either.
+			//
+			// DO NOT accept requests on the new protocol. Otherwise:
+			// 1. We'll end up in V2 routing tables.
+			// 2. We'll have V1 peers in our routing table.
+			//
+			// In other words, we'll pollute the V2 network.
+			queryProtocols = []protocol.ID{cfg.protocolPrefix + kad1}
+			serverProtocols = queryProtocols
+			routingTableProtocols = []string{
+				string(cfg.protocolPrefix + kad2),
```

Comment:
Why do we need to add v2 peers into our routing tables? v2 peers all support the v1 protocol and will respond to requests using it. By adding v1 peers to our routing tables don't we also implicitly add v2 peers?

Reply:
We do, I just have this for completeness. In case we stop advertising v1 in the future.

Reply:
But if in the future we stop advertising v1 then these compat nodes won't be able to send queries to anyone since multistream negotiation will fail right?

Reply:
Uh, well, hm. Yes. I did that so DHT queries could skip the negotiation step and use the "fast" negotiation protocol.

Ok, I'll remove this.

Reply:
If you think it's important/useful we could allow the compat nodes to query using v2 also (i.e. check for v1 and then v2) and that should probably work.

Reply:
Eh, I'd rather not support two query protocols unless we need to. That'll cost us one to two round trips per request.

Reply:
Makes sense. That means this PR can be made really simple then right, since almost no changes (basically just set protocols = server protocols = kad1)

Reply:
I've switched back.